Coconut PR Process fix + Minor update

TwiggyJrr opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Closes https://github.com/BeeStation/BeeStation-Hornet/issues/12559

Processing a coconut in hand will now produce the coconut flesh and the coconut cup instead of just deleting the object.

Coconut process now uses the variable `get_turf()` defined to "location"

- The amount of coconut flesh obtained during processing is based on how much potency the coconut has (between 1 - 6 parts of flesh)
- Chems are now distributed between flesh with coconut milk changing into nutrition and vitamins (/2 and /4 respectivly)
- Removed coconut milk from standard coconut flesh (only able to get via the cup)
- Fixed a bug with densified chemicals trait causing runtime errors
- The coconut cup unit volume size now scales from: 10u (minimum) to 100u (maximum)
    - If the coconut has the densified chemicals trait, the volume size of the cup is doubled up to 200u at 100 potency

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Bug fix = good
New features added to the coconut cup allow for larger beakers in botany that don't _need_ to use glass from the lathe.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.
-->

## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->
<details>
<summary>Screenshots&Videos</summary>

Video showing processing in hand:
https://github.com/user-attachments/assets/0372fee2-24d4-49ea-aca3-b10cc99046bf

Screenshot showing the container size (in the chem dispenser) and the reagents from coconut flesh (in the scanned beaker):
![image](https://github.com/user-attachments/assets/9f2c5c37-58af-4ca7-aed7-d49128c61bf1)


</details>

## Changelog
:cl: Wikimody, TwiggyJrr
fix: Processing a coconut in hand will now produce the coconut flesh & cup.
fix: Fixed a bug with densified chemicals trait causing runtime errors
add: Coconut cup volume size scales on potency, min 10u max 100u. It can go to 200u if it has densified chems trait
add: Coconut flesh now scales on potency size and also contains the plants reagents
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

TwiggyJrr committed:
Coconut process bug fix

Coconuts if processed in hand no longer delete the coconut and give none of the items from the coconut.

TwiggyJrr committed:
Merge branch 'BeeStation:master' into Coconut-PR-Fix

EvilDragonfiend reviewed (changes requested):
use `get_turf(user)`

TwiggyJrr committed:
updated to use get_turf(user)

Updated from (user.loc) to (get_turf(user))

TwiggyJrr committed:
Merge branch 'Coconut-PR-Fix' of https://github.com/TwiggyJrr/BeeStation-Hornet into Coconut-PR-Fix

EvilDragonfiend reviewed (changes requested):
get_turf() should be stored to a variable rather than calling the proc multiple times.

TwiggyJrr committed:
Updated Proc + Cup Scale Size by Wikimody

Defined the get_turf() proc to location to no longer call the proc multiple times.

Updated code from Wikimody:
Now amount of coconuts flesh is based on how much potency coconut has (1 - 6)
Chems are now distributed between parts with coconut milk changing into nutrition and vitamins (/2 and /4 respectively)
Removed coconut milk from standard coconut flesh
Fixed a bug with densified chemicals trait causing runtimes
If densified chemicals is a trait coconut has the size of the cup is doubled up to 200 at 100 potency

TwiggyJrr committed:
Deleted spacing and respaced

Hopefully fixes tabbed spacing?

EvilDragonfiend commented on code/modules/hydroponics/grown/coconut.dm:
```diff
@@ -34,16 +34,52 @@ when processed, it lets you choose between coconut flesh or the coconut cup*/
+		if(seed.get_gene(/datum/plant_gene/trait/maxchem)) // If plant has densified chemicals trait
+			cup.volume = max(20, round(seed.potency) * 2) // Scale volume, minimum of 20 - max of 200
+		else
+			cup.volume = max(10, round(seed.potency)) // Scale volume, minimum of 10 - max of 100
```
You're repeating the logic.
And I don't like the current volume size. Half it, 200u to 100u, 100u to 50u.

wikimody replied:
This is needed for the cup to work with desyfied chemicals. Max chemicals in plant is 100, with densyfied 200. If cup has less volume than that there will be chemical loss where it shoudnt be.

EvilDragonfiend replied:
I am saying line 71 is not effective.

EvilDragonfiend replied:
var/modifier = 1
if(seed.get_gene(/datum/plant_gene/trait/maxchem))
     modifier = 2
cup.volume = max(10, round(seed.potency)) * modifier

this logic is a thing too

EvilDragonfiend replied:
I still do not like the volume size but okay

EvilDragonfiend commented on code/modules/hydroponics/grown/coconut.dm:
```diff
@@ -34,16 +34,52 @@ when processed, it lets you choose between coconut flesh or the coconut cup*/
+	// Defaults to creating 1 coconut flesh when processed
+	var/part_amount = 1
+	if(seed && seed.potency)
+		part_amount = floor(seed.potency / 20 + 1) // Min 1, Max 6 at 100 potency
```
These codes are better at line 54 since it's the only part where it's used.

EvilDragonfiend commented on code/modules/hydroponics/grown/coconut.dm:
```diff
@@ -34,16 +34,52 @@ when processed, it lets you choose between coconut flesh or the coconut cup*/
+
+	// Delete the coconut after processing
 	qdel(src)
 	return ..()
```
This is a bit late, but better to address, I suppose.

..() shouldn't be called after `qdel()`. You should do early return without `..()` or do `..()` first before `qdel()`

EvilDragonfiend replied:
, = ,,()
qdel(src)
this order is safe

EvilDragonfiend commented on code/modules/hydroponics/grown/coconut.dm:
```diff
@@ -34,16 +34,52 @@ when processed, it lets you choose between coconut flesh or the coconut cup*/
+		if(seed.get_gene(/datum/plant_gene/trait/maxchem)) // If plant has densified chemicals trait
+			cup.volume = max(20, round(seed.potency) * 2) // Scale volume, minimum of 20 - max of 200
+		else
+			cup.volume = max(10, round(seed.potency)) // Scale volume, minimum of 10 - max of 100
```
there is no reason to use round when potency is not decimal and you don't divide numbers.

EvilDragonfiend commented on code/modules/hydroponics/grown/coconut.dm:
```diff
@@ -34,16 +34,52 @@ when processed, it lets you choose between coconut flesh or the coconut cup*/
+			if(istype(reg, /datum/reagent/consumable/coconutmilk)) // Remove coconut milk
+				revised_regredients.add_reagent(/datum/reagent/consumable/nutriment/vitamin, reg.volume / 4)
+				revised_regredients.add_reagent(/datum/reagent/consumable/nutriment, reg.volume / 2)
+				revised_regredients.del_reagent(/datum/reagent/consumable/coconutmilk)
```
I think there is a proc something like `has_reagent(/datum/regaent/thing)` afaik, so that you can put into an if condition.
also, the nutriment should be added by the current volume of the coconut milk reagent, not this manual addition.

wikimody replied:
It is added based on coconut milk volume, im not sure what you mean

EvilDragonfiend replied:
check the charcoal code. there is a code using 'has reagent' proc

wikimody replied:
Now that I look at it it might not work with has ingredient since I need that regredient volume to. I will have to see what the function does to be sure

wikimody replied:
Managed to get this work

EvilDragonfiend commented on code/modules/hydroponics/grown/coconut.dm:
```diff
@@ -34,16 +34,52 @@ when processed, it lets you choose between coconut flesh or the coconut cup*/
+	for(var/i = 1 to part_amount)
+		var/obj/item/food/coconutflesh/flesh = new /obj/item/food/coconutflesh(location)
+		if(reagents && reagents.total_volume > 0) // If coconut has no chems, flesh gets default ones
+			flesh.reagents.clear_reagents()
```
rather than creating and clear regents for every item, just make a subtype of coconuetflesh as /empty

wikimody replied:
Its good in case if we got 0 potency plant, it gives us just default

EvilDragonfiend replied:
zero potency shouldn't make anything

EvilDragonfiend commented on code/modules/hydroponics/grown/coconut.dm:
```diff
@@ -34,16 +34,52 @@ when processed, it lets you choose between coconut flesh or the coconut cup*/
+	// Defaults to creating 1 coconut flesh when processed
+	var/part_amount = 1
+	if(seed && seed.potency)
+		part_amount = floor(seed.potency / 20 + 1) // Min 1, Max 6 at 100 potency
```
I don't think `floor()` is correctly used here. did you want to use round()?

wikimody replied:
Its supposed to be floor so only 100 potency gets 6 parts

EvilDragonfiend replied:
ah I mislooked this as clamp(). my mistake

EvilDragonfiend commented on code/modules/hydroponics/grown/coconut.dm:
```diff
@@ -34,16 +34,52 @@ when processed, it lets you choose between coconut flesh or the coconut cup*/
+			if(part_amount == 6) // So 100 potency isn't a punishment
+				revised_regredients.copy_to(flesh.reagents, reagents.total_volume / 5)
+			else
+				revised_regredients.copy_to(flesh.reagents, reagents.total_volume / part_amount)
```
feels very janky

wikimody replied:
And i didn't have better idea than that. This works

EvilDragonfiend replied:
```suggestion
			revised_regredients.copy_to(flesh.reagents, clamp(reagents.total_volume / part_amount, 1, 5))
```
You can store clamp() result into a variable if you see wrapped argument is not readible well

wikimody replied:
Clap shoudl be around part amount not the entire thing but i see how it woudl work

EvilDragonfiend replied:
yes you are right it should be `\ clamp(part_amount, 1, 5)`. Reviewing in mobile is bit limiting my visibility

wikimody committed:
Merge pull request #3 from TwiggyJrr/Coconut-PR-Fix

Coconut pr fix

wikimody committed:
First changes

wikimody committed:
Small fixes

TwiggyJrr committed:
Merge pull request #2 from wikimody/Cookonut-2

Cookonut 2

EvilDragonfiend commented on code/modules/hydroponics/grown/coconut.dm:
```diff
@@ -27,23 +27,56 @@ when processed, it lets you choose between coconut flesh or the coconut cup*/
 	w_class = WEIGHT_CLASS_NORMAL
 	throw_speed = 2
 	throw_range = 4
+	var/max_volume = 100
```
unused line

wikimody replied:
It isn't 
That's what's preventing the runtime with desified hems

TwiggyJrr replied:
Spoke to Wiki regarding this, they said that it is there to prevent the runtimes from densified chemicals.

EvilDragonfiend replied:
> It isn't That's what's preventing the runtime with desified hems

where?

wikimody replied:
> where?

Can we talk on discord, this isn't the best platform for communicating

wikimody replied:
But short version is the desnsified chemicals try to acces max volume variable and cocokonut do to weird inheritance doesn't have it. I made a artificial one here so densyfied chemicals can acces it instead and doesn't trow runtimes

wikimody replied:
I also believe it should work for chem interactions, since they probably use it to. They just probably check if it exists before trying to acces it

TwiggyJrr replied:
Updated growninedbile.dm to have the max_volume variable, the same as food/grown contains.

EvilDragonfiend commented on code/modules/hydroponics/grown/coconut.dm:
```diff
@@ -27,23 +27,56 @@ when processed, it lets you choose between coconut flesh or the coconut cup*/
+	for(var/i = 1 to part_amount)
+		var/obj/item/food/coconutflesh/flesh = new /obj/item/food/coconutflesh/empty(location)
+		if(reagents && reagents.total_volume > 0)
+			revised_regredients.copy_to(flesh.reagents, reagents.total_volume / clamp(part_amount, 1, 5))// So 100 potency isn't a punishment
```
```suggestion
	var/part_amount = 1
	var/div_mod = 1
	if(seed && seed.potency)
		part_amount = floor(seed.potency / 20 + 1) // Min 1, Max 6 at 100 potency
		div_mod = clamp(part_amount, 1, 5) // So 100 potency isn't a punishment

	for(var/i = 1 to part_amount)
		var/obj/item/food/coconutflesh/flesh = new /obj/item/food/coconutflesh/empty(location)
		if(reagents && reagents.total_volume > 0)
			revised_regredients.copy_to(flesh.reagents, reagents.total_volume / div_mod)
```
calculating the number before the loop code is better

EvilDragonfiend commented on code/modules/hydroponics/grown/coconut.dm:
```diff
@@ -27,23 +27,56 @@ when processed, it lets you choose between coconut flesh or the coconut cup*/
+			modifier = 2
+		cup.volume = max(10, seed.potency) * modifier // Without trait 10-100, with it 20-200
+	else
+		cup.volume = 50 // Default volume if potency is unavailable
```
```suggestion
```
I think it is redundant because default volume in the code is already 50u

EvilDragonfiend commented on code/modules/hydroponics/grown/coconut.dm:
a bit nitpick, but I think `reagents` should be `src.reagents` to prevent misreading.
if you are not in favour of using `src` because it has a bad readability too, you can make a variable holder like `var/datum/reagents/original_reagent_holder = reagents`

TwiggyJrr replied:
Updated to use `original_reagent_holder` for readability. Will be making a new commit soon!

EvilDragonfiend reviewed (changes requested):
almost okay

TwiggyJrr committed:
Update code/modules/hydroponics/grown/coconut.dm

Co-authored-by: EvilDragonfiend <87972842+EvilDragonfiend@users.noreply.github.com>

TwiggyJrr committed:
Update code/modules/hydroponics/grown/coconut.dm

Co-authored-by: EvilDragonfiend <87972842+EvilDragonfiend@users.noreply.github.com>

TwiggyJrr committed:
Original_reagent_holder

Copy pastad the code from one file to another to just help my own readability when it came to adding that variable because my brain was getting overwhelmed for some reason (: Apologies for every line changing, I am still learning, lmao

TwiggyJrr committed:
Merge remote-tracking branch 'upstream/master' into Coconut-PR-Fix

TwiggyJrr committed:
adds max_volume to growninedible

Adds the max_volume variable to growninedbile.dm - food/grown has the same variable.

TwiggyJrr committed:
Fix Indentation

Fixed indentation, removed var/max_volume so it doesn't redeclare the variable.

TwiggyJrr committed:
remove max_volume from coconut

Removal of `max_volume` from coconut as it's now declared and handled in the inedible.dm file for inedible plants

EvilDragonfiend commented on code/modules/hydroponics/grown/coconut.dm:
```diff
@@ -34,16 +34,51 @@ when processed, it lets you choose between coconut flesh or the coconut cup*/
+			var/datum/reagent/reg = original_reagent_holder.get_reagent(/datum/reagent/consumable/coconutmilk)
+			revised_regredients.del_reagent(/datum/reagent/consumable/coconutmilk)
+			revised_regredients.add_reagent(/datum/reagent/consumable/nutriment/vitamin, reg.volume / 4)
+			revised_regredients.add_reagent(/datum/reagent/consumable/nutriment, reg.volume / 2)
```
I doubt if `reg.volume` still exists when you do after `del_reagent(/datum/reagent/consumable/coconutmilk)`
please confirm this is okay

wikimody replied:
We are using regredients volume from coconut chems not the revised list used for flesh. The deleting milk first is to make sure no container size shenanigans acur

wikimody replied:
To explain again bu hopefully better
We got 2 lists coconut chems with will end up in cup and revised list with will end up in flesh
We take reference for milk in the coconut chems and delete all milk in the flesh list, then add the nutrients and vitamins based on volume of the unaffected coconut chems milk

EvilDragonfiend replied:
ah, I can recognise it now.
would you rename it to `reagents_template` or something? the var name is a little bit confusing
I will merge once this is addressed.

TwiggyJrr replied:
updated `revised_regredients` to `reagents_template`

EvilDragonfiend commented on code/modules/hydroponics/grown/coconut.dm:
```diff
@@ -34,16 +34,51 @@ when processed, it lets you choose between coconut flesh or the coconut cup*/
+	var/datum/reagents/revised_regredients = new /datum/reagents // Reagents for coconut flesh
+	original_reagent_holder.copy_to(revised_regredients, original_reagent_holder.total_volume)
+
+	if(original_reagent_holder && original_reagent_holder.total_volume > 0)
```
```suggestion
	var/datum/reagents/revised_regredients
	if(original_reagent_holder && original_reagent_holder.total_volume > 0)
		revised_regredients = new /datum/reagents // Reagents for coconut flesh
		original_reagent_holder.copy_to(revised_regredients, original_reagent_holder.total_volume)
```
do not do `new` unless it's necessary

EvilDragonfiend commented on code/modules/hydroponics/grown/coconut.dm:
```diff
@@ -34,16 +34,51 @@ when processed, it lets you choose between coconut flesh or the coconut cup*/
+		original_reagent_holder.trans_to(cup.reagents, original_reagent_holder.total_volume)
+
+	// Delete the coconut after processing
+	. = ..()
```
```suggestion
	. = ..()
	qdel(revised_regredients)
```
should qdel the datum

wikimody replied:
Doesn't it get auto deleted anyway during the whole coconut delete? And if not garbage collector should take care of it

EvilDragonfiend reviewed (changes requested):
seems to be almost alright

TwiggyJrr committed:
Update code/modules/hydroponics/grown/coconut.dm

Co-authored-by: EvilDragonfiend <87972842+EvilDragonfiend@users.noreply.github.com>

TwiggyJrr committed:
Update code/modules/hydroponics/grown/coconut.dm

Co-authored-by: EvilDragonfiend <87972842+EvilDragonfiend@users.noreply.github.com>

TwiggyJrr committed:
updated variable

updated variable from `revised_regredients` to `reagents_template`

TwiggyJrr committed:
Merge remote-tracking branch 'upstream/master' into Coconut-PR-Fix

EvilDragonfiend commented on code/modules/hydroponics/growninedible.dm:
```diff
@@ -9,6 +9,7 @@
 	resistance_flags = FLAMMABLE
 	var/obj/item/seeds/seed = null // type path, gets converted to item on New(). It's safe to assume it's always a seed item.
 	var/discovery_points = 0 //Amount of discovery points given for scanning
+	var/max_volume = 100 // There is the same variable in the food/grown.dm
```
```suggestion
	var/max_volume = 100 // There is the same variable in the food/grown.dm - this variable only exists to suppress a runtime error by /datum/plant_gene/trait/maxchem touching max_volume
```

EvilDragonfiend committed:
too nitpick to ask for a change request
